Foundation/FileManager.swift

```diff
@@ -564,104 +562,268 @@ open class FileManager : NSObject {
+                if !shouldProceedAfterError(error, linkingItemAtPath: srcPath, toPath: dstPath, isURL: isURL) {
+                    throw error
+                }
             }
         }
     }
-
-    open func removeItem(atPath path: String) throws {
+    
+    @inline(__always)
+    func shouldProceedAfterError(_ error: Error, removingItemAtPath path: String, isURL: Bool) -> Bool {
+        guard let delegate = self.delegate else { return false }
+        if isURL {
+            return delegate.fileManager(self, shouldProceedAfterError: error, removingItemAt: URL(fileURLWithPath: path))
+        } else {
+            return delegate.fileManager(self, shouldProceedAfterError: error, removingItemAtPath: path)
+        }
+    }
+    
+    @inline(__always)
+    func shouldRemoveItemAtPath(_ path: String, isURL: Bool) -> Bool {
+        guard let delegate = self.delegate else { return true }
+        if isURL {
+            return delegate.fileManager(self, shouldRemoveItemAt: URL(fileURLWithPath: path))
+        } else {
+            return delegate.fileManager(self, shouldRemoveItemAtPath: path)
+        }
+    }
+    
+    fileprivate func _removeItem(atPath path: String, isURL: Bool, alreadyConfirmed: Bool = false) throws {
+        guard alreadyConfirmed || shouldRemoveItemAtPath(path, isURL: isURL) else {
+            return
+        }
+        
         if rmdir(path) == 0 {
             return
         } else if errno == ENOTEMPTY {
 
-            let fsRep = FileManager.default.fileSystemRepresentation(withPath: path)
-            let ps = UnsafeMutablePointer<UnsafeMutablePointer<Int8>?>.allocate(capacity: 2)
-            ps.initialize(to: UnsafeMutablePointer(mutating: fsRep))
-            ps.advanced(by: 1).initialize(to: nil)
-            let stream = fts_open(ps, FTS_PHYSICAL | FTS_XDEV | FTS_NOCHDIR, nil)
-            ps.deinitialize(count: 2)
-            ps.deallocate()
-            fsRep.deallocate()
-
+            let stream = URL(fileURLWithPath: path).withUnsafeFileSystemRepresentation { (fsRep) -> UnsafeMutablePointer<FTS>? in
+                let ps = UnsafeMutablePointer<UnsafeMutablePointer<Int8>?>.allocate(capacity: 2)
+                ps.initialize(to: UnsafeMutablePointer(mutating: fsRep))
+                ps.advanced(by: 1).initialize(to: nil)
+                defer {
+                    ps.deinitialize(count: 2)
+                    ps.deallocate()
+                }
+                return fts_open(ps, FTS_PHYSICAL | FTS_XDEV | FTS_NOCHDIR, nil)
+            }
+            
             if stream != nil {
                 defer {
                     fts_close(stream)
                 }
 
-                var current = fts_read(stream)
-                while current != nil {
-                    switch Int32(current!.pointee.fts_info) {
+                while let current = fts_read(stream)?.pointee {
+                    
+                    
+                    let itemPath = NSString(bytes: current.fts_path,
```

Comment:
@spevans Can I replace this with `String.init(cString:)`?

Reply:
Best to use `self.string(withFileSystemRepresentation:length:)` to convert to a `String`.